python/tk_framework_desktopserver/server_protocol.py

```diff
@@ -129,56 +181,156 @@ def onMessage(self, payload, is_binary):
         reactor.callInThread(
             self._process_message,
             message_host,
             message,
             message["protocol_version"],
         )
 
+    def _validate_user(self, host_network, user_id, origin_network):
+        """
+        Validates if the user from the browser can connect to this server.
+
+        :returns: True if the user can connect, False otherwise.
+        """
+        # If the hosts are different or the user ids are different, report an error.
+        if host_network != origin_network or (user_id is not None and user_id != self.factory.user_id):
+            logger.debug("Browser integration request received a different user.")
+            logger.debug("Desktop site: %s", host_network)
+            logger.debug("Desktop user: %s", self.factory.user_id)
+            logger.debug("Origin site: %s", origin_network)
+            logger.debug("Origin user: %s", user_id)
+            return False
+        else:
+            return True
+
+    def _handle_get_protocol_version(self):
+        """
+        Handles get_protocol_version
+
+        Note: This is mock by the browser integration crash tool, do not change or you will break
+        the tool.
+        """
+        self.json_reply(dict(protocol_version=self._protocol_version))
+
+    def _is_using_encryption(self):
+        """
+        Checks if this connection requires encryption.
+
+        .. note::
+            While the connection requires encryption, it doesn't necessarily mean that encryption
+            is active on the connection. The initial handshake between the client and the server
+            when requesting for the server id is unencrypted.
+
+        :returns: True if this connection will require encryption.
+        """
+        return True if self.factory.ws_server_id else False
+
+    def _handle_get_ws_server_id(self, message):
+        """
+        Handles the request for the server id.
+
+        This method will retrieve the websocket server secret and send back the server id to the
+        webapp.
+        """
+
+        if not self._is_using_encryption():
+            logger.error(self.ENCRYPTION_NOT_SUPPORTED[1])
+            self.sendClose(*self.ENCRYPTION_NOT_SUPPORTED)
+            return
+
+        # Build a response for the web app.
+        message = Message(message["id"], self._protocol_version)
+        message.reply({
+            "ws_server_id": self.factory.ws_server_id
+        })
+
+        # send the response.
+        self.json_reply(message.data)
+
+        # Create a Fernet instance so we can start encrypting and decrypting messages from now on.
+        self._fernet = Fernet(self._retrieve_server_secret())
+
+    def _retrieve_server_secret(self):
+        """
+        Retrieves the server secret from Shotgun.
+
+        .. note::
+            This method caches the result of the retrieval for any other ServerProtocol instances
+            created after the first one.
+
+            Because
```

Comment:
Good docs, but worried about formatting when it's converted to Sphinx. For each bullet point I would do a a proper sentence with punctuation. Also would check what the "Because" will look like without any other treatment.

Reply:
Good point. I discovered [this tool](http://rst.ninjs.org/#ICAgICAgICAuLiBub3RlOjoKICAgICAgICAgICAgVGhpcyBtZXRob2QgY2FjaGVzIHRoZSByZXN1bHQgb2YgdGhlIHJldHJpZXZhbCBmb3IgYW55IG90aGVyIFNlcnZlclByb3RvY29sIGluc3RhbmNlcwogICAgICAgICAgICBjcmVhdGVkIGFmdGVyIHRoZSBmaXJzdCBvbmUuCgogICAgICAgICAgICBDb25zaWRlciB0aGUgZm9sbG93aW5nCgogICAgICAgICAgICAtIEVhY2ggc2VydmVyIGluc3RhbmNlIGhhcyBhIGRpZmZlcmVudCBzZXJ2ZXIgaWQKICAgICAgICAgICAgLSBSZXRyaWV2aW5nIGEgc2VjcmV0IHdpdGggYSBkaWZmZXJlbnQgaWQgZ2VuZXJhdGVzIGEgbmV3IHNlY3JldAogICAgICAgICAgICAtIFdlIGNhbiBsYXVuY2ggdHdvIHNlcnZlcnMgYXQgdGhlIHNhbWUgdGltZS4KCiAgICAgICAgICAgIEl0IHdvdWxkbid0IGJlIGEgZ29vZCBpZGVhIHRvIHJldHJpZXZlIHRoZSBzZXJ2ZXIgaWQgYmVmb3JlIHN1Y2Nlc3NmdWxseSBsaXN0ZW5pbmcgb24KICAgICAgICAgICAgdGhlIHBvcnQgc2luY2UgbGF1bmNoaW5nIHRoZSBzZXJ2ZXIgYSBzZWNvbmQgdGltZSBieSBtaXN0YWtlIHdvdWxkIHJlc2V0IHRoZSBzZWNyZXQKICAgICAgICAgICAgcmV0cmlldmVkIGJ5IHRoZSBmaXJzdCBpbnN0YW5jZS4=), pretty cool